### PR TITLE
Allowlist for GitLab

### PR DESCRIPTION
**Pull request by mfocko**
TODO:
- [x] DB migration to prefix current entries
- [ ] automatically add GitLab namespaces to waiting list
  - [x] thinking about it: **fix the bug when adding from GitHub installations**
- [x] Agree on the implementation
- [x] Decide if we should support `AllowlistStatus.denied`
- [x] Fix tests and adjust to new behaviour
- [x] Implement decided interface
- [X] Check format of events to prevent ambiguity, e.g. `https://github.com/mfocko/dotfiles.git.git` (ofc GitLab allows it)
   - [x] Shall we save it in the format `github.com/packit/packit` or `github.com/packit/packit.git`? I prefer `.git`, since on GitLab you could get your repository approved and then create subnamespace O.o
   - URLs in hooks don't contain `.git` suffix
- [x] UX of allowlisting new accounts (especially if we add hostname and support separate repositories)

Signed-off-by: Matej Focko <mfocko@redhat.com>

### CODE REVIEW

**Review by TomasTomecek — COMMENTED** [review]
I'm not sure what exactly is your thought-process but I wouldn't personally do it with regexes.

**Comment by TomasTomecek** [comment]
I'd personally do it like this:
```
class AllowlistModel(Base):
    __tablename__ = "allowlist"
    id = Column(Integer, primary_key=True)

    # this should be the account name of the person who installed packit in github - not sure what would the gitlab equiv would be
    account_name = Column(String, index=True)

    # ideally we should also store the fas account name of the requester
    fas_account = Column(String)

    # string or enum
    forge = {github.com, gitlab.com}

    # root namespace, let's just approve everything for users or namespaces - I don't see what the benefit would be for being able to block subnamespaces
    namespace = Column(String)
    status = Column(Enum(AllowlistStatus))
```

and translate the existing data in the schema migration

is this clear, does it make sense?

**Comment by lachmanfrantisek** [comment]
> since on GitLab you could get your repository approved and then create subnamespace O.o

If they want to be evil, they can do evil things in the already-approved repo as well.

**Comment by mfocko** [comment]
based on Tomas' suggestion I propose:
```py
class AllowlistModel(Base):
    __tablename__ = "allowlist"
    id = Column(Integer, primary_key=True)
    namespace = Column(String, index=True) # previously account_name

    parent_id = Column(Integer, ForeignKey("allowlist.id"), nullable=True)
    parent = relationship("AllowlistModel")

    # parent = relationship("AllowlistModel", back_populates="subnamespaces")
    # subnamespaces = relationship("AllowlistModel", back_populates="parent")
 
    status = Column(Enum(AllowlistStatus))
    # ideally we should also store the fas account name of the requester
    fas_account = Column(String)
```

I would probably approach it as Trie(s)
- it would allow us to store github/gitlab only once (with implicit `AllowlistStatus.Deny`)
- we can differentiate forges, e.g. `github.com/some_namespace` is allowed and we get triggered on `gitlab.com/some_namespace/some_repo.git`, we will find `some_namespace` entry, but parent is `github.com` => denied
- FAS username is fine (related to plans for implementation of automatic approval based on `packager` group and/or CLA)

**Comment by TomasTomecek** [comment]
I'd say the parent mechanism is overkill but won't stop you if you think it's a good idea :) since you'll be writing the code :D 

On the other hand I agree this will give us complete flexibility and explicit permissions for every single repo in a namespace - I just wanna make sure:
1. we'll have code to allow all repos in a namespace (and all subnamespaces)
2. p-s will work the same way for current users - meaning, they will not notice we have changed in how allowlist works

**Comment by mfocko** [comment]
> I'd say the parent mechanism is overkill but won't stop you if you think it's a good idea :) since you'll be writing the code :D

"A bit" :D also Franta has suggested we should hit around 3 levels at most, so majority of entries will be namespaces like are now, they would just keep one link to "root namespace" = github/gitlab (even allowing self-hosted instances of gitlab nicely)

> 1. we'll have code to allow all repos in a namespace (and all subnamespaces)

That's the intention... Let's say we allowlist `gitlab.com/packit-service/src`,
- we get request for `gitlab.com/packit-service/src/glibc.git`
  1. can't find `glibc.git`, let's go one level up
  2. found `src` (allowed), let's check rest of the path, done
- we get request for `github.com/src/something.git`
  1. can't find `something.git`, level up
  2. found `src` (allowed), let's check rest of the path, from database i get `packit-service`, but I have `github.com` => denied

it mostly hits UX of the allowlist script we use; will have to adjust it too

> 2. p-s will work the same way for current users - meaning, they will not notice we have changed in how allowlist works

yeah, apart from allowlisting namespaces of some bots and developers yeah :D (that's relict of allowing namespace OR user; could be probably cleaned up against projects from which we received webhooks)

**Comment by TomasTomecek** [comment]
@mfocko got it now, thanks for the detailed explanation, makes complete sense to me, very nice design!

**Comment by mfocko** [comment]
updated the tests, will probably create fake DB in fixture, since it looks awful right now; there are added comments also, the format is `(queried_namespace, returned_db_entries)`, in that order (as described [here](https://github.com/packit/packit-service/pull/1001#issuecomment-786076251))

File: tests/unit/test_allowlist.py [hunk 1]
```diff
@@ -46,39 +46,301 @@ def allowlist():
...
+                                status=AllowlistStatus.denied,
+                                parent_id=8,
+                                parent=flexmock(
+                                    id=8,
+                                    namespace="gitlab.com",
+                                    status=AllowlistStatus.denied,
+                                    parent_id=None,
+                                ),
+                            ),
+                        )
```
Comment (mfocko) [inline tests/unit/test_allowlist.py:238]: missing github.com with implicit deny...

File: tests/unit/test_allowlist.py [hunk 2]
```diff
@@ -46,39 +46,301 @@ def allowlist():
...
+                            parent=flexmock(
+                                id=1,
+                                account_name="github.com",
+                                status=AllowlistStatus.denied,
+                                parent_id=None,
+                            ),
+                        )
+                    ),
+                )
+            ),
```
Comment (mfocko) [inline tests/unit/test_allowlist.py:135]: missing gitlab.com

**Comment by TomasTomecek** [comment]
@mfocko would you have time to finish the work this week?

**Comment by mfocko** [comment]
> @mfocko would you have time to finish the work this week?

I hope so, majority is done, so card can be closed imo
we're talking about:
- database migration
- script for approving that we use internally
- handling of `waiting` since we can handle the repositories themselves

**Comment by mfocko** [comment]
Please go by commits when reviewing. In case I rebased and squashed something I didn't want, **YELL**. :D

File: alembic/versions/800abbbb23c9_allow_multiple_forges_in_allowlist.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+    session.commit()
+
+
+def downgrade():
+    bind = op.get_bind()
+    session = orm.Session(bind=bind)
+
+    for entry in session.query(AllowlistModel).all():
+        if entry.namespace.startswith("github.com/"):
```
Comment (mfocko) [inline alembic/versions/800abbbb23c9_allow_multiple_forges_in_allowlist.py:95]: this way we can keep the gitlab accounts when downgrading, they are unusable with the previous implementation... could be probably a better idea to check for domain when upgrading too :thinking:

File: alembic/versions/800abbbb23c9_allow_multiple_forges_in_allowlist.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        )
+
+
+def upgrade():
+    op.add_column("allowlist", sa.Column("fas_account", sa.String(), nullable=True))
+
+    # rename account_name to namespace
+    op.execute("ALTER TABLE allowlist RENAME COLUMN account_name TO namespace")
+
+    # fix index; will it work with just rename? O.o
```
Comment (TomasTomecek) [inline alembic/versions/800abbbb23c9_allow_multiple_forges_in_allowlist.py:75]: you can try locally with stg data :)

File: alembic/versions/800abbbb23c9_allow_multiple_forges_in_allowlist.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    op.execute("ALTER INDEX ix_allowlist_account_name RENAME TO ix_allowlist_namespace")
+    # op.create_index(op.f('ix_allowlist_namespace'), 'allowlist', ['namespace'], unique=False)
+    # op.drop_index('ix_allowlist_account_name', table_name='allowlist')
+
+    # update all of the entries
+    bind = op.get_bind()
+    session = orm.Session(bind=bind)
+
+    for entry in session.query(AllowlistModel).all():
+        entry.namespace = f"github.com/{entry.namespace}"
```
Comment (TomasTomecek) [inline alembic/versions/800abbbb23c9_allow_multiple_forges_in_allowlist.py:85]: so one of the first things we'll need to do will be to approve gitlab.com/packit-service after we deploy this to stg

File: tests/unit/test_allowlist.py [hunk 6]
```diff
@@ -46,39 +46,151 @@ def allowlist():
...
+            id=8,
+            namespace="gitlab.com/packit",
+            status=AllowlistStatus.approved_manually.value,
+        ),
+        "github.com/packit": flexmock(
+            id=10,
+            namespace="github.com/packit",
+            status=AllowlistStatus.denied.value,
+        ),
+        "gitlab.com/packit-service/src": flexmock(
```
Comment (TomasTomecek) [inline tests/unit/test_allowlist.py:88]: could we get another layer for gitlab here? e.g. `gitlab.com/packit-service/src/systemd`

File: tests/unit/test_allowlist.py [hunk 7]
```diff
@@ -46,39 +46,151 @@ def allowlist():
...
+            ("github.com/konipas", "github.com"),
+            False,
+        ),
+        (  # Krasomila is not present at all, checks parent
+            "github.com/krasomila",
+            ("github.com/krasomila", "github.com"),
+            False,
+        ),
+        (  # gitlab.com/packit-service/src is approved
+            "gitlab.com/packit-service/src/glibc.git",
```
Comment (TomasTomecek) [inline tests/unit/test_allowlist.py:163]: oh, you're checking it here

File: tests/unit/test_allowlist.py (hunk 6 above)
Comment (TomasTomecek) [reply tests/unit/test_allowlist.py:88]: ...so what should happen here actually for gitlab.com?

**Review by TomasTomecek — COMMENTED** [review]
Awesome! Really love that you added way-more tests and updated docs and the script as well.

File: alembic/versions/800abbbb23c9_allow_multiple_forges_in_allowlist.py (hunk 3 above)
Comment (TomasTomecek) [reply alembic/versions/800abbbb23c9_allow_multiple_forges_in_allowlist.py:95]: afaik for gitlab it's being used only in gitlab.com/packit-service (and soon gitlab.com/redhat/centos-stream/src)

File: alembic/versions/800abbbb23c9_allow_multiple_forges_in_allowlist.py (hunk 4 above)
Comment (mfocko) [reply alembic/versions/800abbbb23c9_allow_multiple_forges_in_allowlist.py:75]: already tried with prod data :D seems to be fine... we probably also could set the FAS accounts based on `automatically_approved`, but it's just matched login, which would go against the already open issue for actually verifying through the login to FAS.

I'm still thinking about adding the option to keep `gitlab.com/` prefixes when downgrading, just in case

File: tests/unit/test_allowlist.py (hunk 6 above)
Comment (mfocko) [reply tests/unit/test_allowlist.py:88]: > could we get another layer for gitlab here? e.g. `gitlab.com/packit-service/src/systemd`

test? sure

> ...so what should happen here actually for gitlab.com?

nothing, it just looks it up as parent namespace that can be approved; this could be potentially helpful with self-hosted instances with restricted access (e.g. for testing), it would allow allowlisting whole instance

File: tests/unit/test_allowlist.py (hunk 7 above)
Comment (mfocko) [reply tests/unit/test_allowlist.py:163]: oh, I interpreted it as another subnamespace in the comment above

File: tests/unit/test_allowlist.py (hunk 6 above)
Comment (TomasTomecek) [reply tests/unit/test_allowlist.py:88]: yup, as a test case

thanks for explanation

File: tests/unit/test_allowlist.py (hunk 7 above)
Comment (TomasTomecek) [reply tests/unit/test_allowlist.py:163]: I just wanted to be sure the behaviour is correct above

File: alembic/versions/800abbbb23c9_allow_multiple_forges_in_allowlist.py (hunk 4 above)
Comment (lachmanfrantisek) [inline alembic/versions/800abbbb23c9_allow_multiple_forges_in_allowlist.py:75]: If this is tested, let's remove that comment...;)

File: alembic/versions/800abbbb23c9_allow_multiple_forges_in_allowlist.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+def upgrade():
+    op.add_column("allowlist", sa.Column("fas_account", sa.String(), nullable=True))
+
+    # rename account_name to namespace
+    op.execute("ALTER TABLE allowlist RENAME COLUMN account_name TO namespace")
+
+    # fix index; will it work with just rename? O.o
+    op.execute("ALTER INDEX ix_allowlist_account_name RENAME TO ix_allowlist_namespace")
+    # op.create_index(op.f('ix_allowlist_namespace'), 'allowlist', ['namespace'], unique=False)
+    # op.drop_index('ix_allowlist_account_name', table_name='allowlist')
```
Comment (lachmanfrantisek) [inline alembic/versions/800abbbb23c9_allow_multiple_forges_in_allowlist.py:78]: Why is this here?

**Review by lachmanfrantisek — COMMENTED** [review]
Thanks for all the work!

Please, update the migration script (only comments?) so it looks like it will really work..;)

File: alembic/versions/800abbbb23c9_allow_multiple_forges_in_allowlist.py (hunk 8 above)
Comment (mfocko) [reply alembic/versions/800abbbb23c9_allow_multiple_forges_in_allowlist.py:78]: alembic initially generated `create + drop` for the index, i checked after the migration (with the `rename` that is uncommented):
```
packit=# \d allowlist 
                                    Table "public.allowlist"
   Column    |       Type        | Collation | Nullable |                Default                
-------------+-------------------+-----------+----------+---------------------------------------
 id          | integer           |           | not null | nextval('allowlist_id_seq'::regclass)
 namespace   | character varying |           |          | 
 status      | allowliststatus   |           |          | 
 fas_account | character varying |           |          | 
Indexes:
    "allowlist_pkey" PRIMARY KEY, btree (id)
    "ix_allowlist_namespace" btree (namespace)

packit=# \d ix_allowlist_namespace 
     Index "public.ix_allowlist_namespace"
    Column    |       Type        | Definition 
--------------+-------------------+------------
 account_name | character varying | namespace
btree, for table "public.allowlist"
```

so it is OK even if it keeps the old column name in index, right? cause it registers `ix_*` as index from the table

Comment (lachmanfrantisek) [reply alembic/versions/800abbbb23c9_allow_multiple_forges_in_allowlist.py:78]: Aha, so we can remove those lines or put a comment that the rename replaces create&drop.

**Review by lachmanfrantisek — APPROVED** [review]
:rocket:
